src/ngx_http_lua_kong_module.c

```diff
@@ -624,3 +713,562 @@ ngx_http_lua_kong_get_upstream_ssl_verify(ngx_http_request_t *r,
+
+    b->last = ngx_copy(b->last, method.data, method.len);
+    *b->last++ = ' ';
+
+    u->uri.data = b->last;
+
+    if (plcf->proxy_lengths && ctx->vars.uri.len) {
+        b->last = ngx_copy(b->last, ctx->vars.uri.data, ctx->vars.uri.len);
+
+    } else if (unparsed_uri) {
+        b->last = ngx_copy(b->last, r->unparsed_uri.data, r->unparsed_uri.len);
+
+    } else {
+        if (r->valid_location) {
+            b->last = ngx_copy(b->last, ctx->vars.uri.data, ctx->vars.uri.len);
+        }
+
+        if (escape) {
+            ngx_escape_uri(b->last, r->uri.data + loc_len,
+                           r->uri.len - loc_len, NGX_ESCAPE_URI);
+            b->last += r->uri.len - loc_len + escape;
+
+        } else {
+            b->last = ngx_copy(b->last, r->uri.data + loc_len,
+                               r->uri.len - loc_len);
+        }
+
+        if (r->args.len > 0) {
+            *b->last++ = '?';
+            b->last = ngx_copy(b->last, r->args.data, r->args.len);
+        }
+    }
+
+    u->uri.len = b->last - u->uri.data;
+
+    if (plcf->http_version == NGX_HTTP_VERSION_11) {
+        b->last = ngx_cpymem(b->last, ngx_http_proxy_version_11,
+                             sizeof(ngx_http_proxy_version_11) - 1);
+
+    } else {
+        b->last = ngx_cpymem(b->last, ngx_http_proxy_version,
+                             sizeof(ngx_http_proxy_version) - 1);
+    }
+
+    ngx_memzero(&e, sizeof(ngx_http_script_engine_t));
+
+    e.ip = headers->values->elts;
+    e.pos = b->last;
+    e.request = r;
+    e.flushed = 1;
+    le.ip = headers->lengths->elts;
+
+    while (*(uintptr_t *) le.ip) {
+
+        lcode = *(ngx_http_script_len_code_pt *) le.ip;
+        (void) lcode(&le);
+
+        for (val_len = 0; *(uintptr_t *) le.ip; val_len += lcode(&le)) {
+            lcode = *(ngx_http_script_len_code_pt *) le.ip;
+        }
+        le.ip += sizeof(uintptr_t);
+
+        if (val_len == 0) {
+            e.skip = 1;
+
+            while (*(uintptr_t *) e.ip) {
+                code = *(ngx_http_script_code_pt *) e.ip;
+                code((ngx_http_script_engine_t *) &e);
+            }
+            e.ip += sizeof(uintptr_t);
+
+            e.skip = 0;
+
+            continue;
+        }
+
+        code = *(ngx_http_script_code_pt *) e.ip;
+        code((ngx_http_script_engine_t *) &e);
+
+        *e.pos++ = ':'; *e.pos++ = ' ';
+
+        if (ngx_strstr(e.pos - ngx_strlen("Host: "), "Host: ") != NULL &&
+            r->headers_in.host->value.data != NULL) {
+            // keep the end of the Host header name position
+            orig_pos = e.pos;
+            // run the script
+            while (*(uintptr_t *) e.ip) {
+                code = *(ngx_http_script_code_pt *) e.ip;
+                code((ngx_http_script_engine_t *) &e);
+            }
+            // overwrite the Host header value
+            e.pos = orig_pos;
+            e.pos = ngx_copy(e.pos, r->headers_in.host->value.data, r->headers_in.host->value.len);
```

Comment:
if i understand correctly, `e.pos` is point to buffer `b` which is created from Line 1054 with the length calculated from known headers. if we overwrite the header larger than the previous, this is going to overflow. (let's also make the test a HTTP 1.0 request in test to force nginx to parse all CRLF correctly and validate all sent headers are intact)

Reply:
Yes, the length need to be calculated based on the actual header size, instead of the original headers.